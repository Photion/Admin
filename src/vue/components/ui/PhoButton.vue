<template>
  <component :is="attrs.is" v-bind="attrs">
    <slot>
      <FontAwesomeIcon v-if="icon" :icon="icon" />
      <span v-if="label">{{ label }}</span>
    </slot>
  </component>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue';

import { componentProps, useId } from '%/shared';
import FontAwesomeIcon from '%/ui/forms/FontAwesomeIcon.vue';

export default defineComponent({
  components: {
    FontAwesomeIcon,
  },
  props: {
    ...componentProps,
    color: {
      type: String,
      default: () => 'primary',
      validator: (color: string) => {
        return ['primary', 'secondary', 'success', 'danger', 'disabled'].indexOf(color) !== -1;
      },
    },
    href: {
      type: String,
      default: () => '',
    },
    icon: {
      type: String,
      default: () => '',
    },
    label: {
      type: String,
      default: () => '',
    },
  },
  setup(props) {
    const id = useId('button', props);

    const attrs = computed(() => {
      const values: Record<string, unknown> = {
        is: computed(() => props.href ? 'a' : 'button'),
        id: id.unique,
        cy: id.family,
        class: `pho-btn ${props.color}`,
      };

      if (props.color === 'disabled') {
        values.disabled = 'disabled';
      }

      if (props.href) {
        const url = new URL(props.href, window.location.origin);

        if (url.origin === window.location.origin) {
          values.is = 'router-link';
          values.to = props.href;
        } else {
          values.is = 'a';
          values.href = props.href;
        }
      } else {
        values.is = 'button';
        values.type = 'button';
      }

      return values;
    });

    return {
      attrs,
    };
  },
});
</script>
