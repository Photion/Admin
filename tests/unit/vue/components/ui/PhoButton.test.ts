import { mount } from '@vue/test-utils';

import PhoButton from '~/src/vue/components/ui/PhoButton.vue';

const getComponent = <T>(props: T) => {
  return mount(PhoButton, {
    props: {
      name: 'pho-button-comp',
      uuid: 'my-uuid',
      ...props,
    },
  });
};

describe('unit.vue.components.ui.PhoButton', () => {

  it('Mounts', () => {
    const wrapper = getComponent({});
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Sets the attributes', () => {
    const wrapper = getComponent({});
    const attributes = wrapper.get('[cy]').attributes();

    expect(attributes).toMatchSnapshot();
  });

  it('Renders the label', async () => {
    const wrapper = getComponent({ label: 'Magic Button' });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Renders the icon', async () => {
    const wrapper = getComponent({ icon: 'times' });
    expect(wrapper.element).toMatchSnapshot();
  });

});
