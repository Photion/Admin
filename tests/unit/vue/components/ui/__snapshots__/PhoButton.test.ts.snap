// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.vue.components.ui.PhoButton Mounts 1`] = `
<button
  class="pho-btn primary"
  cy="button:pho-button-comp"
  id="button:pho-button-comp:my-uuid"
  is="button"
  type="button"
>
  
  <!--v-if-->
  <!--v-if-->
  
</button>
`;

exports[`unit.vue.components.ui.PhoButton Renders a remote link 1`] = `
<a
  class="pho-btn primary"
  cy="button:pho-button-comp"
  href="https://developer.mozilla.org/en-US/docs/Web/API/URL/URL"
  id="button:pho-button-comp:my-uuid"
  is="a"
>
  
  <!--v-if-->
  <!--v-if-->
  
</a>
`;

exports[`unit.vue.components.ui.PhoButton Renders an internal link 1`] = `
<router-link
  class="pho-btn primary"
  cy="button:pho-button-comp"
  id="button:pho-button-comp:my-uuid"
  is="router-link"
  to="/abc"
>
  
  <!--v-if-->
  <!--v-if-->
  
</router-link>
`;

exports[`unit.vue.components.ui.PhoButton Renders the icon 1`] = `
<button
  class="pho-btn primary"
  cy="button:pho-button-comp"
  id="button:pho-button-comp:my-uuid"
  is="button"
  type="button"
>
  
  <svg
    class=""
    style="width: 12px;"
    viewBox="0 0 352 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
      fill="currentColor"
    />
  </svg>
  <!--v-if-->
  
</button>
`;

exports[`unit.vue.components.ui.PhoButton Renders the label 1`] = `
<button
  class="pho-btn primary"
  cy="button:pho-button-comp"
  id="button:pho-button-comp:my-uuid"
  is="button"
  type="button"
>
  
  <!--v-if-->
  <span>
    Magic Button
  </span>
  
</button>
`;

exports[`unit.vue.components.ui.PhoButton Sets the attributes 1`] = `
Object {
  "class": "pho-btn primary",
  "cy": "button:pho-button-comp",
  "id": "button:pho-button-comp:my-uuid",
  "is": "button",
  "type": "button",
}
`;
